languages/python/sqlalchemy-oso/sqlalchemy_oso/session.py

```diff
@@ -210,12 +213,56 @@ class AuthorizedSession(AuthorizedSessionBase, Session):
     Usually :py:func:`authorized_sessionmaker` is used instead of directly
     instantiating the session.
 
-    NOTE: Unless ``enable_baked_queries=True`` is passed to the constructor,
-          _baked_queries are disabled since the caching mechanism can bypass
-          authorization by using queries from the cache that were previously
-          baked without authorization applied.
+    NOTE: _baked_queries are disabled on SQLAlchemy 1.3 since the caching
+          mechanism can bypass authorization by using queries from the cache
+          that were previously baked without authorization applied. Note that
+          _baked_queries are deprecated as of SQLAlchemy 1.4.
 
-    .. _baked_queries: https://docs.sqlalchemy.org/en/13/orm/extensions/baked.html
+    .. _baked_queries: https://docs.sqlalchemy.org/en/14/orm/extensions/baked.html
     """
 
     pass
+
+
+if AT_LEAST_SQLALCHEMY_VERSION_1_4:
+    # TODO(gj): remove type ignore once we upgrade to 1.4-aware MyPy types.
+    from sqlalchemy.orm import with_loader_criteria  # type: ignore
+
+    @event.listens_for(Session, "do_orm_execute")
+    def do_orm_execute(execute_state):
+        if not execute_state.is_select:
+            return
+
+        session = execute_state.session
+
+        if not isinstance(session, AuthorizedSessionBase):
+            return
+        assert isinstance(session, Session)
+
+        oso: Oso = session.oso_context["oso"]
+        user = session.oso_context["user"]
+        action = session.oso_context["action"]
+
+        def entities_in_statement(statement):
+            def _entities_in_statement(statement):
+                try:
+                    entities = (cd["entity"] for cd in statement.column_descriptions)
+                    return set(e for e in entities if e is not None)
+                except AttributeError:
+                    return set()
+
+            entities = _entities_in_statement(statement)
+
+            # TODO(gj): currently walking way more than we have to. Probably
+            # some points in the tree where we can safely call it good for that
+            # branch and continue on to more fruitful pastures.
+            for child in statement.get_children():
+                entities |= entities_in_statement(child)
+
+            return entities
+
+        for entity in entities_in_statement(execute_state.statement):
+            filter = authorize_model(oso, user, action, session, entity)
+            if filter is not None:
+                where = with_loader_criteria(entity, filter, include_aliases=True)
```

Comment:
This is cool! Do you happen to know if this works correctly for lazy relationship loads? 

I'm asking because of this bug in SQLAlchemy 1.3: https://www.notion.so/osohq/SQLAlchemy-Authorization-bypass-with-joined-eager-loader-cea9dd6d256c4930a8ef33cc0385fdce

(If you don't no need to investigate we can just add a note)

Reply:
Not sure